# Fundamentos de Aprendizaje Automatico
## Pratica 1
### Alfonso Bonilla Trueba y Monica de la Iglesia Martinez

##### Apartado 1 (Usando nuestra implementacion de Naive Bayes)

* <b>Ejecuciones german.data primero sin correción y despues con correción</b>

In [1]:
%run TasasClasificador.py ./ConjuntosDatos/german.data

  Fichero:  german.data
  - Porcentaje Entrenamiento en Validacion simple 70.00%
  - Numero de Particiones en Validacion cruzada:  5
  - Sin usar Correccion de La Place
  - Usando nuestra libreria
----------------------------------------------------
  El error con Validacion Simple es del 24.33%
  El error medio con Validacion Cruzada es del 23.70% Con una desviacion de 3.19%


In [2]:
%run TasasClasificador.py ./ConjuntosDatos/german.data --laPlace

  Fichero:  german.data
  - Porcentaje Entrenamiento en Validacion simple 70.00%
  - Numero de Particiones en Validacion cruzada:  5
  - Usando la Correccion de La Place
  - Usando nuestra libreria
----------------------------------------------------
  El error con Validacion Simple es del 29.67%
  El error medio con Validacion Cruzada es del 23.40% Con una desviacion de 2.42%


|           | <font size="5"><b>german.data</b></font>|        |
| :-------: | :------: | :-----: |
|           | <font size="4"><b>Validación Simple</b></font> | <font size="4"><b>Validación Cruzada</b></font> |
| <font size="3">Sin Corrección</font>  | <font size="3">24.33%</font>  |  <font size="3">23.70%</font> |
| <font size="3">Con Corección</font>  | <font size="3">29.67%</font> |    <font size="3">23.40%</font> |

* <b>Ejecuciones tic-tac-toe.data primero sin correción y despues con correción</b>

In [3]:
%run TasasClasificador.py ./ConjuntosDatos/tic-tac-toe.data

  Fichero:  tic-tac-toe.data
  - Porcentaje Entrenamiento en Validacion simple 70.00%
  - Numero de Particiones en Validacion cruzada:  5
  - Sin usar Correccion de La Place
  - Usando nuestra libreria
----------------------------------------------------
  El error con Validacion Simple es del 30.90%
  El error medio con Validacion Cruzada es del 28.33% Con una desviacion de 2.18%


In [11]:
%run TasasClasificador.py ./ConjuntosDatos/tic-tac-toe.data --laPlace

  Fichero:  tic-tac-toe.data
  - Porcentaje Entrenamiento en Validacion simple 70.00%
  - Numero de Particiones en Validacion cruzada:  5
  - Usando la Correccion de La Place
  - Usando nuestra libreria
----------------------------------------------------
  El error con Validacion Simple es del 29.51%
  El error medio con Validacion Cruzada es del 29.58% Con una desviacion de 2.07%


|           | <font size="5"><b>tic-tac-toe.data</b></font>|        |
| :-------: | :------: | :-----: |
|           | <font size="4"><b>Validación Simple</b></font> | <font size="4"><b>Validación Cruzada</b></font> |
| <font size="3">Sin Corrección</font>  | <font size="3">30.90%</font>  |  <font size="3">28.33%</font> |
| <font size="3">Con Corección</font>  | <font size="3">29.51%</font> |    <font size="3">29.58%</font> |

##### Apartado 2

Vamos a comenzar comparando los resulados para el fichero german.data. Usando correccion de La place vemos que tanto en validación simple como en validación cruzada el porcentaje de error ha aumentado. Esto ocurre para los casos en los que en las tablas de ocurrencia hay algun cero sumamos uno a toda esa tabla. En este caso el 0 no es por falta de datos puesto que es un fichero grande, sino que es porque ese caso no se da. Por tanto al aplicar la correcion estamos falseando los datos y obteniendo predicciones erroneas en algun caso. Este es el motivo por el que hay un ligero aumento de la tasa de error usando La Place.

Para el fichero tic-tac-toe.data vemos que el error es practicamente el mimso, es decir, las tablas de ocurrencia usadas para el calculo de Naive Bayes no hay casi 0, por tanto, la correccion no tiene mucho efecto.

##### Apartado 3 (Usando la libreria scikit-learn)

* <b>Ejecuciones german.data primero sin correción y despues con correción</b>

In [6]:
%run TasasClasificador.py ./ConjuntosDatos/german.data --sklearn

  Fichero:  german.data
  - Porcentaje Entrenamiento en Validacion simple 70.00%
  - Numero de Particiones en Validacion cruzada:  5
  - Sin usar Correccion de La Place
  - Usando la libreria Sklearn
----------------------------------------------------
  El error con Validacion Simple es del 28.24%
  El error con Validacion Cruzada es del 27.50% Con una desviacion de 3.19%


In [7]:
%run TasasClasificador.py ./ConjuntosDatos/german.data --sklearn --laPlace

  Fichero:  german.data
  - Porcentaje Entrenamiento en Validacion simple 70.00%
  - Numero de Particiones en Validacion cruzada:  5
  - Usando la Correccion de La Place
  - Usando la libreria Sklearn
----------------------------------------------------
  El error con Validacion Simple es del 36.88%
  El error con Validacion Cruzada es del 36.70% Con una desviacion de 4.08%


|           | <font size="5"><b>german.data</b></font>|        |
| :-------: | :------: | :-----: |
|           | <font size="4"><b>Validación Simple</b></font> | <font size="4"><b>Validación Cruzada</b></font> |
| <font size="3">Sin Corrección</font>  | <font size="3">28.24%</font>  |  <font size="3">27.50%</font> |
| <font size="3">Con Corección</font>  | <font size="3">36.88%</font> |    <font size="3">36.70%</font> |

* <b>Ejecuciones tic-tac-toe.data primero sin correción y despues con correción</b>

In [8]:
%run TasasClasificador.py ./ConjuntosDatos/tic-tac-toe.data --sklearn

  Fichero:  tic-tac-toe.data
  - Porcentaje Entrenamiento en Validacion simple 70.00%
  - Numero de Particiones en Validacion cruzada:  5
  - Sin usar Correccion de La Place
  - Usando la libreria Sklearn
----------------------------------------------------
  El error con Validacion Simple es del 34.03%
  El error con Validacion Cruzada es del 32.99% Con una desviacion de 2.22%


In [9]:
%run TasasClasificador.py ./ConjuntosDatos/tic-tac-toe.data --sklearn --laPlace

  Fichero:  tic-tac-toe.data
  - Porcentaje Entrenamiento en Validacion simple 70.00%
  - Numero de Particiones en Validacion cruzada:  5
  - Usando la Correccion de La Place
  - Usando la libreria Sklearn
----------------------------------------------------
  El error con Validacion Simple es del 25.00%
  El error con Validacion Cruzada es del 30.37% Con una desviacion de 3.58%


|           | <font size="5"><b>tic-tac-toe.data</b></font>|        |
| :-------: | :------: | :-----: |
|           | <font size="4"><b>Validación Simple</b></font> | <font size="4"><b>Validación Cruzada</b></font> |
| <font size="3">Sin Corrección</font>  | <font size="3">34.03%</font>  |  <font size="3">32.99%</font> |
| <font size="3">Con Corección</font>  | <font size="3">25.00%</font> |    <font size="3">30.37%</font> |

#### Comparacion de nuestra implementacion con scikit-learn

Podemos observar que los errores usando scikit-learn son mayores con el fichero german. Esto es porque este fichero contiene atributos continuos. Estos atributos son soportados por nuestra implementación, pero no lo scikit-learn. Por tanto a la hora de calcular las probabilidades nuestro clasificador es mas preciso.

El fichero tic-tac-toe no contiene atributos continuos, por lo que la tasa de error de nuestro clasificador y la de scikit-learn son mas aproximadas. Es imposible que salgan igual, puesto que antes de hacer las particiones se permutan los datos para evitar cualquier tipo de sesgo. Por lo cual, no estan entrenando y testeando el mismo conjunto de datos, aunque estos provengan del mismo fichero de datos.





##### Apartado 4

El inconveniente que encontramos en la validación cruzada es encontrar un buen k (numero de particiones). Este depende de la cantidad de datos que se tengan. Si tienes pocos datos pero eliges un k muy alto, el error medio obtenido será muy elevado, puesto que el tamaño de las particiones es muy pequeño, y por tanto el test es tan pequeño obtiene una tasa de fallo muy elevada.

Computacionalmente tambien creemos que es mas costoso, puesto que la validacion cruzada es como realizar k validaciones simples.

##### Información adicional

Hemos implementado un main denominado TasasClasificador.py
Este programa recibe como minimo la ruta del fichero que contiene los datos por argumento.

> ~$ python TasasClasificador.py ./ConjuntosDatos/german.data

Este main que hemos implementado soporta los siguientes argumentos opcionales:
* --laPlace                 Ejecuta el Clasificador con corrección La Place
* --sklearn                 Ejecuta el Clasificador NB de la libreria scikit-learn
* --porcenaje <0.0>         Establece el porcentaje de train
* --particiones <0>         Número entero que establece el numero de particiones para la validación cruzada


Si no se introduce ninguno de estos argumentos opcionales por defecto se ejecuta el clasificador implementado por nosotros, sin correción La Place, con un porcentaje de entrenamiento del 70% para validación simple y 5 particiones para validación cruzada.